quinn-proto/src/crypto/rustls.rs

```diff
@@ -0,0 +1,384 @@
+    pub fn new(config: rustls::ClientConfig) -> Self {
+        Self(Arc::new(config))
+    }
+
+    /// Add a trusted certificate authority.
+    ///
+    /// For more advanced/less secure certificate verification, construct a [`ClientConfig`]
+    /// manually and use rustls's `dangerous_configuration` feature to override the certificate
+    /// verifier.
+    pub fn add_certificate_authority(&mut self, cert: Certificate) -> Result<(), webpki::Error> {
+        let anchor = webpki::trust_anchor_util::cert_der_as_trust_anchor(untrusted::Input::from(
+            &cert.inner.0,
+        ))?;
+        Arc::make_mut(&mut self.0)
+            .root_store
+            .add_server_trust_anchors(&webpki::TLSServerTrustAnchors(&[anchor]));
+        Ok(())
+    }
+
+    /// Enable NSS-compatible cryptographic key logging to the `SSLKEYLOGFILE` environment variable
+    ///
+    /// Useful for debugging encrypted communications with protocol analyzers such as Wireshark.
+    pub fn enable_keylog(&mut self) {
+        Arc::make_mut(&mut self.0).key_log = Arc::new(KeyLogFile::new());
+    }
+
+    /// Set the application-layer protocols to accept, in order of descending preference
+    ///
+    /// When set, clients which don't declare support for at least one of the supplied protocols will be rejected.
+    ///
+    /// The IANA maintains a [registry] of standard protocol IDs, but custom IDs may be used as well.
+    ///
+    /// [registry]: https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml#alpn-protocol-ids
+    pub fn set_protocols(&mut self, protocols: &[&[u8]]) {
+        Arc::make_mut(&mut self.0).alpn_protocols = protocols.iter().map(|x| x.to_vec()).collect();
+    }
+}
+
+impl Default for ClientConfig {
+    fn default() -> ClientConfig {
+        let mut cfg = rustls::ClientConfig::new();
+        cfg.versions = vec![ProtocolVersion::TLSv1_3];
+        cfg.enable_early_data = true;
+        Self(Arc::new(cfg))
+    }
+}
+
+impl Deref for ClientConfig {
+    type Target = Arc<rustls::ClientConfig>;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for ClientConfig {
+    fn deref_mut(&mut self) -> &mut Arc<rustls::ClientConfig> {
+        &mut self.0
+    }
+}
+
+impl crypto::ClientConfig<TlsSession> for ClientConfig {
+    fn start_session(
+        &self,
+        server_name: &str,
+        params: &TransportParameters,
+    ) -> Result<TlsSession, ConnectError> {
+        let pki_server_name = DNSNameRef::try_from_ascii_str(server_name)
+            .map_err(|_| ConnectError::InvalidDnsName(server_name.into()))?;
+        Ok(TlsSession::Client(rustls::ClientSession::new_quic(
+            &self.0,
+            pki_server_name,
+            to_vec(params),
+        )))
+    }
+}
+
+/// rustls configuration for server sessions
+#[derive(Clone)]
+pub struct ServerConfig(Arc<rustls::ServerConfig>);
+
+impl ServerConfig {
```

Comment:
We should expose a way to construct this from a `rustls::ServerConfig`, as was done with with `ClientConfig`

Reply:
Done. Can you reapprove? (I think your Gitter bridge might have failed?)